File: array and string/diagonal_traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        if(matrix.length==0 || matrix==null)
            return new int[0];
        
        int[] arr = new int[matrix.length*matrix[0].length];
        int r=0;
        int c=0;
        int i=0;
        while(i<arr.length)
        {
            // Move up the diagonal
            while(i<arr.length && r>=0 && c<matrix[0].length)
            {
                arr[i++] = matrix[r--][c++];
            }
            r++;
            
            // if column number has reached maximum
            if(c==matrix[0].length)
            {
                r++;
                c--;
            }
            
            // Move down rhe diagonal
            while(i<arr.length && r<matrix.length && c>=0)
            {
                arr[i++] = matrix[r++][c--];
            }
            c++;
            
            // if row number has reached maximum
            if(r==matrix.length)
            {
                r--;
                c++;
            }
            
        }
        
        return arr;
    }
}

File: array and string/diagonal_traverse.py
class Solution:
    def findDiagonalOrder(self, matrix: List[List[int]]) -> List[int]:
        if not(matrix):
            return []
        i=0
        arr = []
        r=0
        c=0
        
        while i<len(matrix)*len(matrix[0]):
            
            # Move up the diagonal
            while(i<len(matrix)*len(matrix[0]) and r>=0 and c<len(matrix[0])):
                arr.append(matrix[r][c])
                r-=1
                c+=1
                i+=1
            r+=1
            
            # if column reached its maximum length
            if c==len(matrix[0]):
                c-=1
                r+=1
            
            # Move down the diagonal
            while(i<len(matrix)*len(matrix[0]) and r<len(matrix) and c>=0):
                arr.append(matrix[r][c])
                r+=1
                c-=1
                i+=1
            c+=1
            
            # if row reached its maximum length
            if r==len(matrix):
                c+=1
                r-=1
        
        return arr